File: arrays_hashing/validAnagram.cpp
#include <unordered_map>
#include <iostream>
#include <string>
using namespace std;


class Solution {
public:
    bool isAnagram(string s, string t) {
        unordered_map<char,int> sMap,tMap;
        
        if(s.size()!= t.size()){
            return false;
        }
        for(int i = 0; i<s.size(); i++){
            if(sMap.count(s[i])) sMap[s[i]]++;
            else sMap[s[i]] = 1;

            if(tMap.count(t[i])) tMap[t[i]]++;
            else tMap[t[i]] = 1;
        }

        for(const auto& [key,val] : sMap){
            if( !tMap.count(key) || tMap[key] != val) return false;
        }
        return true;
    }
};


int main(){
    Solution s;
    cout << s.isAnagram("anagram","nagaram") << endl;
    cout << s.isAnagram("rat","car") << endl;
    return 0;
}

File: arrays_hashing/validAnagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        sMap = {}
        tMap = {}
        sLen = len(s)
        tLen = len(t)
        if sLen != tLen:
            return False
        
        for i in range(sLen):
            sChar = s[i]
            tChar = t[i]

            if sChar not in sMap: sMap[sChar] = 1
            else: sMap[sChar] +=1

            if tChar not in tMap: tMap[tChar] = 1
            else: tMap[tChar] +=1
        
        for key,val in sMap.items():
            if key not in tMap or tMap[key] != val:
                return False

        return True
    

def main():
    s = Solution()
    print(s.isAnagram("anagram", "nagaram"))
    print(s.isAnagram("rat", "car"))

if __name__ == "__main__":
    main()